lib/tpm2_session.c

```diff
@@ -164,3 +172,150 @@ tpm2_session *tpm2_session_new(TSS2_SYS_CONTEXT *sapi_context,
 
     return session;
 }
+
+#define SESSION_VERSION 1
+
+/*
+ * Checks that two types are equal in size.
+ *
+ * It works by leveraging the fact that C does not allow negative array sizes.
+ * If the sizes are equal, the boolean equality operator will return 1, thus
+ * a subtraction of 1 yields 0, which is a legal array size in C. In the false
+ * case (ie sizes not equal), 0 - 1 is -1, which will cause the compiler to
+ * complain.
+ */
+#define COMPILE_ASSERT_SIZE(a, b) \
+    typedef char WRONG_SIZE_##a[(sizeof(a) == sizeof(b)) - 1]
+
+// We check that the TSS library does not change sizes unbeknownst to us.
+COMPILE_ASSERT_SIZE(TPM2_HANDLE, UINT32);
+COMPILE_ASSERT_SIZE(TPMI_ALG_HASH, UINT16);
+COMPILE_ASSERT_SIZE(TPM2_SE, UINT8);
```

Comment:
Hmm, I wonder if were aren't being too paranoid here. At the end of the day we depend on the TSS library to not break API and ABI compatibility, otherwise shouldn't we check the size of every possible type?

Reply:
Because it will get silently truncated and debugging it will be a PITA.
Ie this generates no warnings on GCC with no other options enabled:
```
    uint32_t a = 2^10;
    uint8_t x = a;
```